File: mcm/2differentWaysToAddParenthesis.py
class Solution:
    def __init__(self):
        self.dp = {}
    def diffWaysToCompute(self, expression):
        if expression in self.dp:
            return self.dp[expression]
        else:
            result = []
            num = 0
            for i in range(len(expression)):
                c = expression[i]
                if c != '*' and c != '-' and c != '+':
                    num = num * 10 + int(c)
                else:
                    if expression[: i] in self.dp:
                        left = self.dp[expression[: i]]
                    else:
                        left = self.diffWaysToCompute(expression[: i])
                    if expression[i + 1:] in self.dp:
                        right = self.dp[expression[i + 1:]]
                    else:
                        right = self.diffWaysToCompute(expression[i + 1:])
                    for l in left:
                        for r in right:
                            if c == '*':
                                result.append(l * r)
                            elif c == '+':
                                result.append(l + r)
                            else:
                                result.append(l - r)
            if len(result) == 0:
                result.append(num)
                self.dp[expression] = result
            return result

File: mcm/2differentWaysToAddParenthesis.cpp
class Solution {
    unordered_map<string, vector<int>> seen;
public:
    vector<int> diffWaysToCompute(string expression){
        if (seen.count(expression)) return seen[expression];
        // we have only single number
        vector<int> result, left, right;
        int num = 0;
        for (int i = 0; i < expression.size(); i++){
            char ch = expression[i];
            if (ch != '+' && ch != '-' && ch != '*'){
                num = (num * 10) + ch - '0';
            }
            else {
                left = diffWaysToCompute(expression.substr(0, i));
                right = diffWaysToCompute(expression.substr(i + 1));
            }
            for (auto l: left){
                for (auto r: right){
                    if (ch == '+') result.push_back(l + r);
                    else if (ch == '*') result.push_back(l * r);
                    else if (ch == '-') result.push_back(l - r);
                }
            }
        }
        // base case
	// single number
        if (result.size() == 0){
            result.push_back(num);
        }
        seen[expression] = result;
        return result;
    }
};
